File: worker/src/teamdetail.js
// teamdetail.js — the RICH tier for a team (the /v1/teamdetail route), mirroring
// the scoreboard-vs-summary split for games: the lean /v1/team card is what the
// home feed polls; this is the one-off, slow-moving detail a team page opens.
// Pure (no I/O), like team.js/normalize.js, so it runs in Node tests + the worker
// alike. The schedule reuses the SAME buildEvent() the scoreboard does, and the
// standing reuses normalizeStandings() — team card, detail, and standings never
// fork. Roster/stats discriminate STRUCTURALLY, never on sport name.

import { resolve } from '../../schema/tools/resolve.mjs';
import { buildEvent, https } from './normalize.js';
import { normalizeStandings } from './standings.js';
import { teamIdentityOf } from './team.js';

const titleCase = (s) =>
  String(s || '').replace(/\b\w/g, (c) => c.toUpperCase());

// ---- roster ------------------------------------------------------------------
// VERIFIED 2026-07: ESPN returns EITHER a flat `athletes[]` (NBA/MLB/NHL) OR a
// grouped `athletes[{position, items:[…]}]` (NFL by offense/defense/specialTeam,
// soccer by position group). Discriminate by the presence of `items[]` — never by
// sport name. Anything unrecognized degrades to one "Roster" group / [].
function mapAthlete(a) {
  const o = {
    id: String(a.id ?? ''),
    name: a.displayName || a.fullName || a.shortName || '',
  };
  if (a.jersey != null && a.jersey !== '') o.jersey = String(a.jersey);
  if (a.position?.abbreviation) o.position = a.position.abbreviation;
  const hs = https(a.headshot?.href || a.headshot);
  if (hs) o.headshot = hs;
  return o;
}

function buildRoster(roster) {
  const athletes = roster?.athletes;
  if (!Array.isArray(athletes) || !athletes.length) return [];
  const grouped = athletes.some((e) => Array.isArray(e.items));
  if (grouped) {
    return athletes
      .filter((g) => Array.isArray(g.items) && g.items.length)
      .map((g) => ({
        name: titleCase(g.position || g.name || 'Group'),
        athletes: g.items.map(mapAthlete).filter((a) => a.id),
      }))
      .filter((g) => g.athletes.length);
  }
  return [{ name: 'Roster', athletes: athletes.map(mapAthlete).filter((a) => a.id) }];
}

// ---- season stats ------------------------------------------------------------
// results.stats.categories[] → TeamStatGroup[]. When the family curates
// `teamStatKeys` (registry), collapse to ONE ordered group of just those keys
// (mirrors standingsColumns curation); else keep the natural categories, capped.
// VERIFIED 2026-07: values arrive as strings/numbers; EPL ships an empty
// `results:{}` in the offseason → [].
function mapStat(s) {
  const value = s.displayValue ?? (s.value != null ? String(s.value) : undefined);
  if (value == null) return null;
  const o = {
    name: s.name || s.abbreviation || '',
    label: s.shortDisplayName || s.displayName || s.name || '',
    value: String(value),
  };
  if (s.abbreviation) o.abbr = s.abbreviation;
  if (typeof s.rank === 'number') o.rank = s.rank;
  return o;
}

function buildStats(profile, stats) {
  const cats = stats?.results?.stats?.categories;
  if (!Array.isArray(cats) || !cats.length) return [];
  const keys = profile.teamStatKeys;
  if (Array.isArray(keys) && keys.length) {
    const byName = {};
    for (const c of cats) for (const s of c.stats || []) {
      if (s?.name && byName[s.name] == null) byName[s.name] = s;
    }
    const picked = keys.map((k) => byName[k]).filter(Boolean).map(mapStat).filter(Boolean);
    return picked.length ? [{ name: 'Season', stats: picked }] : [];
  }
  return cats
    .map((c) => ({
      name: c.displayName || c.name || '',
      stats: (c.stats || []).slice(0, 8).map(mapStat).filter(Boolean),
    }))
    .filter((g) => g.stats.length);
}

// ---- standing (this team's group only) --------------------------------------
// Run the shared standings normalizer, then pluck the group that contains this
// teamId → { groupName, rows }. Omitted when the team isn't found (a national
// team, or an athlete-shaped racing table where no team id matches — but team
// pages are gated on competitorKind==='team' upstream anyway).
function pluckStanding(profile, standingsRaw, teamId) {
  if (!standingsRaw) return undefined;
  const groups = normalizeStandings(standingsRaw);
  const id = String(teamId);
  for (const g of groups) {
    if (g.rows.some((r) => r.team.id === id)) {
      // carry the family's preferred columns (as /v1/standings does) so the team
      // page renders W/L/PCT labels, not raw ESPN stat keys — no extra fetch.
      return { groupName: g.name, columns: profile.standingsColumns || null, rows: g.rows };
    }
  }
  return undefined;
}

// ---- top level ---------------------------------------------------------------
export function normalizeTeamDetail(reg, key, teamId, parts = {}) {
  const { schedule, roster, stats, standingsRaw } = parts;
  const profile = resolve(reg, key);
  const team = teamIdentityOf(profile, schedule?.team, teamId);

  // Full season schedule, normalized through the SAME builder as the scoreboard,
  // start-ascending so the client can slice "last N / next N".
  const events = (schedule?.events || []).map((e) => buildEvent(profile, e));
  events.sort((a, b) => (Date.parse(a.start) || 0) - (Date.parse(b.start) || 0));

  const out = {
    league: key,
    sport: profile.espnSport,
    leagueName: profile.name || key.split('/')[1] || '',
    team,
    schedule: events,
    roster: buildRoster(roster),
    stats: buildStats(profile, stats),
  };
  const standing = pluckStanding(profile, standingsRaw, teamId);
  if (standing) out.standing = standing;
  return out;
}

File: app/lib/src/data/teamdetail.dart
// teamdetail.dart — Dart port of worker/src/teamdetail.js. The RICH tier for a
// team page: full-season schedule + roster + season stats + this team's standings
// group. Reuses buildEvent + normalizeStandings + teamIdentityOf so card, detail,
// and standings never fork. Roster/stats discriminate STRUCTURALLY, never by sport.

import 'profiles.dart';
import 'util.dart';
import 'normalize.dart';
import 'standings.dart';
import 'team.dart';

String _titleCase(dynamic s) =>
    jsStr(s).replaceAllMapped(RegExp(r'\b\w'), (m) => m.group(0)!.toUpperCase());

// ---- roster -----------------------------------------------------------------
Map<String, dynamic> _mapAthlete(dynamic a) {
  final o = <String, dynamic>{
    'id': jsStr(field(a, 'id') ?? ''),
    'name': or([field(a, 'displayName'), field(a, 'fullName'), field(a, 'shortName'), '']),
  };
  if (field(a, 'jersey') != null && field(a, 'jersey') != '') o['jersey'] = jsStr(field(a, 'jersey'));
  if (field(field(a, 'position'), 'abbreviation') != null) o['position'] = field(field(a, 'position'), 'abbreviation');
  final hs = https(or([field(field(a, 'headshot'), 'href'), field(a, 'headshot')]));
  if (hs != null) o['headshot'] = hs;
  return o;
}

List<Map<String, dynamic>> _buildRoster(dynamic roster) {
  final athletes = field(roster, 'athletes');
  if (athletes is! List || athletes.isEmpty) return [];
  final grouped = athletes.any((e) => field(e, 'items') is List);
  if (grouped) {
    return athletes
        .where((g) => field(g, 'items') is List && (field(g, 'items') as List).isNotEmpty)
        .map((g) => {
              'name': _titleCase(or([field(g, 'position'), field(g, 'name'), 'Group'])),
              'athletes': (field(g, 'items') as List).map(_mapAthlete).where((a) => truthy(a['id'])).toList(),
            })
        .where((g) => (g['athletes'] as List).isNotEmpty)
        .toList();
  }
  return [
    {'name': 'Roster', 'athletes': athletes.map(_mapAthlete).where((a) => truthy(a['id'])).toList()}
  ];
}

// ---- season stats -----------------------------------------------------------
Map<String, dynamic>? _mapStat(dynamic s) {
  final value = field(s, 'displayValue') ?? (field(s, 'value') != null ? jsStr(field(s, 'value')) : null);
  if (value == null) return null;
  final o = <String, dynamic>{
    'name': or([field(s, 'name'), field(s, 'abbreviation'), '']),
    'label': or([field(s, 'shortDisplayName'), field(s, 'displayName'), field(s, 'name'), '']),
    'value': jsStr(value),
  };
  if (truthy(field(s, 'abbreviation'))) o['abbr'] = field(s, 'abbreviation');
  if (field(s, 'rank') is num) o['rank'] = field(s, 'rank');
  return o;
}

List<Map<String, dynamic>> _buildStats(Map profile, dynamic stats) {
  final cats = field(field(field(stats, 'results'), 'stats'), 'categories');
  if (cats is! List || cats.isEmpty) return [];
  final keys = profile['teamStatKeys'];
  if (keys is List && keys.isNotEmpty) {
    final byName = <String, dynamic>{};
    for (final c in cats) {
      for (final s in (field(c, 'stats') is List ? field(c, 'stats') as List : const [])) {
        final n = field(s, 'name');
        if (n != null && byName[n] == null) byName[n] = s;
      }
    }
    final picked = keys.map((k) => byName[k]).where((s) => s != null).map(_mapStat).where((s) => s != null).cast<Map<String, dynamic>>().toList();
    return picked.isNotEmpty ? [{'name': 'Season', 'stats': picked}] : [];
  }
  return cats
      .map((c) => {
            'name': or([field(c, 'displayName'), field(c, 'name'), '']),
            'stats': (field(c, 'stats') is List ? (field(c, 'stats') as List).take(8) : const []).map(_mapStat).where((s) => s != null).cast<Map<String, dynamic>>().toList(),
          })
      .where((g) => (g['stats'] as List).isNotEmpty)
      .toList();
}

// ---- standing (this team's group only) --------------------------------------
Map<String, dynamic>? _pluckStanding(Map profile, dynamic standingsRaw, dynamic teamId) {
  if (standingsRaw == null) return null;
  final groups = normalizeStandings(standingsRaw);
  final id = jsStr(teamId);
  for (final g in groups) {
    if ((g['rows'] as List).any((r) => field(field(r, 'team'), 'id') == id)) {
      return {'groupName': g['name'], 'columns': profile['standingsColumns'], 'rows': g['rows']};
    }
  }
  return null;
}

// ---- top level --------------------------------------------------------------
Map<String, dynamic> normalizeTeamDetail(Registry reg, String key, dynamic teamId, [Map parts = const {}]) {
  final schedule = parts['schedule'];
  final roster = parts['roster'];
  final stats = parts['stats'];
  final standingsRaw = parts['standingsRaw'];
  final profile = resolve(reg, key);
  final team = teamIdentityOf(profile, field(schedule, 'team'), teamId);

  final events = (field(schedule, 'events') is List ? field(schedule, 'events') as List : const [])
      .map((e) => buildEvent(profile, e as Map))
      .toList();
  events.sort((a, b) {
    final da = DateTime.tryParse(jsStr(a['start']))?.millisecondsSinceEpoch ?? 0;
    final db = DateTime.tryParse(jsStr(b['start']))?.millisecondsSinceEpoch ?? 0;
    return da.compareTo(db);
  });

  final out = <String, dynamic>{
    'league': key,
    'sport': profile['espnSport'],
    'leagueName': or([profile['name'], key.split('/').length > 1 ? key.split('/')[1] : null, '']),
    'team': team,
    'schedule': events,
    'roster': _buildRoster(roster),
    'stats': _buildStats(profile, stats),
  };
  final standing = _pluckStanding(profile, standingsRaw, teamId);
  if (standing != null) out['standing'] = standing;
  return out;
}
